File: src/merge_sort.cpp
#include <iostream>


#include <vector>

#define eb emplace_back

using namespace std;


vector<int> merge(const vector<int> &l, const vector<int> &r)
{
	vector<int> ans;
	int il=0, ir=0;
	while (il < l.size() && ir < r.size()) {
		if (l[il] <=r[ir]) {
			ans.eb(l[il++]);
		}
		else {
			ans.eb(r[ir++]);
		}
	}
	while (il < l.size()) {
		ans.eb(l[il++]);
	}
	while (ir < r.size()) {
		ans.eb(r[ir++]);
	}
	return ans;
}


vector<int> mergeSort(const vector<int> &a, int l, int r)
{
	vector<int> ans;
	if(l >= r) {
		ans.eb(a[r]);
		return ans;
	}
	int mid = l +(r-l)/2;
	vector<int> la = mergeSort(a, l, mid);
	vector<int> lr = mergeSort(a, mid+1, r);
	ans = merge(la, lr);
	return ans;
}


int main()
{
	vector<int> a = {1, 7, 3, 9, 17, 5, -1};

	vector<int> sorted = mergeSort(a, 0, a.size()-1);
	
	for (auto &i : sorted ) {
		cout << i << endl;
	}
	return 0;

}

File: python/merge_sort.py
def merge(left_arr, right_arr):
    
    res = []
    li = 0
    ri = 0

    while li < len(left_arr) and ri < len(right_arr):
        if left_arr[li] <= right_arr[ri]:
            res.append(left_arr[li])
            li += 1
        else:
            res.append(right_arr[ri])
            ri += 1 
    while li < len(left_arr):
        res.append(left_arr[li])
        li += 1
    while ri < len(right_arr):
        res.append(right_arr[ri])
        ri += 1
    
    return res


def merge_sort(arr):
    if len(arr) <= 1:
        return arr
    mid = len(arr)/2
    res_left = merge_sort(arr[:mid])
    res_right = merge_sort(arr[mid:])
    res = merge(res_left, res_right)
    return res



#res = merge([1, 5, 7, 25], [2, 4, 37, 95])
res = merge_sort([1, -7, 25, 35, 2, 4, 7 ,3, 15])

for i in res:
    print i
